### ISSUE TITLE
Discrepancies between "evaluate" accuracy and "predict" result

### ISSUE THREAD

**Issue by ssoulhi** [fchollet/keras#6421]
Hi All,
I have a  numerical dataset where
X = dataset[:,0:31]
Y = dataset[:,[31,32]]
The model is:
model = Sequential()
model.add(Dense(40, activation='relu',  kernel_initializer='uniform', input_dim=31))
model.add(Dense(31, activation='relu', kernel_initializer='uniform'))
model.add(Dense(2, activation='sigmoid', kernel_initializer='uniform'))

model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['categorical_accuracy'])

 
model.fit(X, Y, epochs=1, batch_size=10,  verbose=2)

At the training the loss is 0.0032 and the categorical_accuracy is 0.8831
After Evaluation with same training data:  score, acc = model.evaluate(X, Y)
the loss is  0.000153304499... and the accuracy is 100%

When running predict: predictions = model.predict(X)
I got very bad results. Below 1st 10 rows of Y and Y_hat:
Y
[[ 0.  0.]
 [ 0.  0.]
 [ 0.  0.]
 [ 0.  0.]
 [ 0.  0.]
 [ 0.  0.]
 [ 0.  0.]
 [ 0.  0.]
 [ 0.  0.]
 [ 0.  0.]]
Y_hat
[[ 1.  0.]
 [ 1.  0.]
 [ 1.  0.]
 [ 1.  0.]
 [ 1.  0.]
 [ 1.  0.]
 [ 1.  0.]
 [ 1.  0.]
 [ 1.  0.]
 [ 1.  0.]]

How come the evaluate accuracy is 100%  while the prediction from model.predict is bad. 

Note that I am using same X and Y for training, evaluate and predict on purpose.

Please advise what I am missing

Thanks for your help

**Comment by bstriner** [comment]
You're using categorical crossentropy and sigmoid output. Two options:
1) Categorical crossentropy with a softmax output layer with two units. Targets are (n,2) one-hot vectors.
2) binary crossentropy with a sigmoid output layer with one unit. Targets are (n,1) 0s or 1s. This only works if there are 2 categories.

Cheers

**Comment by ssoulhi** [comment]
I tried your suggestion #1: model.add(Dense(2, activation='softmax', kernel_initializer='uniform'))
but I have still same issue. Evaluate acc=100% and bad predictions:
Y
[[ 0.  0.]
 [ 0.  0.]
 [ 0.  0.]
 [ 0.  0.]
 [ 0.  0.]
 [ 0.  0.]
 [ 0.  0.]
 [ 0.  0.]
 [ 0.  0.]
 [ 0.  0.]]
Y_hat
[[ 1.  0.]
 [ 1.  0.]
 [ 1.  0.]
 [ 1.  0.]
 [ 1.  0.]
 [ 1.  0.]
 [ 1.  0.]
 [ 1.  0.]
 [ 1.  0.]
 [ 1.  0.]]

**Comment by bstriner** [comment]
But your ys are still 0,0. You were doing categorical crossentropy, which is for making categorical predictions. Like you have 2 buckets and you are trying to predict your bucket. Softmax means the two buckets sum to 1, so it is a valid distribution. `np.sum(Y, axis=0)` should be all 1s if you're using a softmax and categorical crossentropy.

What are you trying to do? What are y0 and y1 supposed to be? For categorical crossentropy each row of your target should be [0,1] or [1,0].

If y0 and y1 are two independent binary predictions, then you should use binary_crossentropy and two sigmoid outputs.

Basically, you have to understand what y is to make a meaningful model.

Cheers

**Comment by ssoulhi** [comment]
With binary_crossentropy the problem is solved. I have now valid predictions. My assumption that categorical_crossentropy models to binary output was wrong.
Thanks a lot for your help !!!!!
Cheers
